### PR TITLE
[MU4] Midievents redesign

### PR DESCRIPTION
**Pull request by igorkorsukov**

### CODE REVIEW

File: mscore/CMakeLists.txt [hunk 1]
```diff
@@ -91,7 +91,7 @@ QT5_WRAP_UI (ui_headers
       synthcontrol.ui splitstaff.ui selectinstr.ui
       editstafftype.ui
       editpitch.ui editstringdata.ui editraster.ui mediadialog.ui albummanager.ui layer.ui realizeharmonydialog.ui
-      omrpanel.ui masterpalette.ui harmonyedit.ui pathlistdialog.ui
```
Comment (vpereverzev) [inline mscore/CMakeLists.txt:94]: I'm not quite sure how it relates to your PR?

File: framework/midi/internal/fluidlitesynth.h [hunk 2]
```diff
@@ -44,15 +44,15 @@ class FluidLiteSynth : public ISynthesizer
     Ret loadSF(const io::path& filePath) override;
     Ret setupChannels(const Programs& programs) override;
 
-    bool handleEvent(uint16_t chan, const Event& e) override;
+    bool handleEvent(const Event& e) override;
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/midi/internal/fluidlitesynth.h:47]: hmm, we are here in the midi context (namespace), in this context there are no other events, so I'm not sure if it is worth renaming. Then to many other types of data, we need to add the midi prefix (for example, channels)

File: framework/midi/internal/sequencer.cpp [hunk 3]
```diff
@@ -93,23 +90,25 @@ void Sequencer::process(float sec)
...
+    uint32_t maxTicks = this->maxTick(m_midiData.events);
     if (m_midiStream->isStreamingAllowed && curTicks >= maxTicks) {
         requestData(curTicks);
     }
 
-    sendEvents(curTicks);
+    sendEvents(prevTicks, curTicks);
 
-    m_lastTimeMsec = msec;
+    m_prevMSec = m_curMSec;
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/midi/internal/sequencer.cpp:111]: Monotonically increasing time is pass to the sequencer (1.1, 1.2, 1.3, 1.4 sec, etc.)
Current - what was passing at the moment, for example, 1.3 sec
Previous - what was previous, for example, 1.2 sec

Accordingly, we process ticks between 1.2 and 1.3 seconds

File: framework/midi/internal/sequencer.cpp [hunk 4]
```diff
@@ -158,8 +156,10 @@ bool Sequencer::run(float init_sec)
         return false;
     }
 
-    m_lastTimeMsec = static_cast<uint64_t>(init_sec * 1000);
-    doRun();
+    m_prevMSec = static_cast<uint64_t>(init_sec * 1000);
+    m_curMSec = m_prevMSec;
+
+    m_internalRunning = true;
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/midi/internal/sequencer.cpp]: Previously it was necessary ... now I looked, I no longer need it, I removed it.

File: framework/midi/internal/sequencer.cpp [hunk 5]
```diff
@@ -168,103 +168,40 @@ bool Sequencer::run(float init_sec)
...
-void Sequencer::reset()
-{
-    m_curMsec = 0;
-    m_seekMsec = 0;
-    for (auto it = m_chanStates.begin(); it != m_chanStates.end(); ++it) {
-        it->second.eventIndex = 0;
-    }
-}
-
-bool Sequencer::doRun()
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/midi/internal/sequencer.cpp:185]: It used to be: 
* run and doRun
* seek and doSeek
* stop and doStop

But now the doX methods have become very small and meaningless, so I removed them and left only the main methods.

File: framework/midi/internal/fluidlitesynth.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (vpereverzev) [reply framework/midi/internal/fluidlitesynth.h:47]: Oh, I hadn't noticed the midi namespace
Agreed

File: framework/midi/internal/sequencer.cpp [hunk 6]
```diff
@@ -304,63 +241,52 @@ void Sequencer::buildTempoMap()
...
+    auto pos = m_midiData.events.lower_bound(fromTick);
+    if (pos == m_midiData.events.end()) {
+        return false;
+    }
 
-    while (1)
-    {
-        if (channelEOT(chan)) {
-            return ret;
+    while (1) {
```
[the remark this answers is not in the document]
Comment (igorkorsukov) [reply framework/midi/internal/sequencer.cpp:260]: Not all events are sent, but only with the set ticks. This is usually a very quick operation.
In the current design, the sequencer is usually used in a separate thread, so there is no problem with blocking it (if it suddenly goes through)

File: mscore/CMakeLists.txt (hunk 1 above)
Comment (igorkorsukov) [reply mscore/CMakeLists.txt:94]: Someone added the `masterpalette.ui` file in this CMakeList in the master, but it is not in the `mscore`, it is in `mu4/scenes/palette`

File: framework/midi/internal/sequencer.cpp [hunk 7]
```diff
@@ -168,103 +167,37 @@ bool Sequencer::run(float init_sec)
...
 }
 
-void Sequencer::doSeek(uint64_t seekMsec)
+void Sequencer::seek(float sec)
 {
-    m_internalRunning = false;
-
-    m_seekMsec = seekMsec;
-    uint32_t seekTicks = ticks(m_seekMsec);
+    uint64_t seekMsec = static_cast<uint64_t>(sec * 1000.f);
```
Comment (psmokotnin) [inline framework/midi/internal/sequencer.cpp:185]: `sec` can be below zero

Comment (psmokotnin) [inline framework/midi/internal/sequencer.cpp:185]: sec can be below zero

Comment (igorkorsukov) [reply framework/midi/internal/sequencer.cpp:189]: fixed, added assert 

```
    IF_ASSERT_FAILED(!(sec < 0)) {
        sec = 0;
    }
```

**Pull request merged**